include/pybind11/attr.h

```diff
@@ -517,23 +517,26 @@ template <size_t Nurse, size_t Patient> struct process_attribute<keep_alive<Nurs
 template <typename... Args> struct process_attributes {
     static void init(const Args&... args, function_record *r) {
         PYBIND11_WORKAROUND_INCORRECT_MSVC_C4100(r);
+        PYBIND11_WORKAROUND_INCORRECT_ALL_GCC_UNUSED_BUT_SET_PARAMETER(r);
         int unused[] = { 0, (process_attribute<typename std::decay<Args>::type>::init(args, r), 0) ... };
-        ignore_unused(unused);
+        silence_unused_warnings(unused);
     }
     static void init(const Args&... args, type_record *r) {
         PYBIND11_WORKAROUND_INCORRECT_MSVC_C4100(r);
+        PYBIND11_WORKAROUND_INCORRECT_ALL_GCC_UNUSED_BUT_SET_PARAMETER(r);
         int unused[] = { 0, (process_attribute<typename std::decay<Args>::type>::init(args, r), 0) ... };
-        ignore_unused(unused);
+        silence_unused_warnings(unused);
     }
     static void precall(function_call &call) {
         PYBIND11_WORKAROUND_INCORRECT_MSVC_C4100(call);
         int unused[] = { 0, (process_attribute<typename std::decay<Args>::type>::precall(call), 0) ... };
```

Comment:
```suggestion
        (int []) { 0, (process_attribute<typename std::decay<Args>::type>::precall(call), 0) ... };
```

Why can't we just do this? I'm guessing the point of this is to "trick" the compile to get it to call all these in an expansion, but we could at least save several lines and forcing the reader to rely on what "unused" and "silence unused warnings" is.

I may not fully understand the trick here though, since I don't understand the `0,` at the start is doing. That's not part of the usual trick I've seen elsewhere.

Reply:
Just quick: that's existing code, the only reason I touched it is to use the more general new function (I'm not wedded to the function name). But I'll play with your idea to see how it pans out.

Reply:
Yes, I know, and I don't mind the name, I just think we should fix the issue rather than hack around with the hack for it. I don't think we want to create a variable here, since we don't use it; making a variable should be telling the reader you are going to do something with it.

Reply:
I naively would put `(int []) {process_attribute<typename std::decay<Args>::type>::precall(call) ... };` - we should at least comment on why this structure is needed if it is.

Reply:
I did a quick experiment:

https://github.com/pybind/pybind11/actions/runs/1086709363

I didn't look in detail, but as far as I can see all MSVC and clang builds break. Only gcc and icc don't issue warnings.

An example clang warning (my local machine) is below. Note that I don't use `-Werror` on my local machine, therefore the build succeeds and the tests run successfully. Your suggested change is clearly correct, but not compatible with MSVC and clang without some sort of warning suppression we'd need to figure out.

@henryiii It's a fair question to ask, but tangential to this PR. I'm already paying a very high price unblocking refactoring and IWYU cleanup by painstakingly nibbling away on the pragma block in pybind11.h. I need to stay focused on that. I could open a bug or another PR with your suggested change, for someone to work on separately. Please let me know if you think that's worthwhile. Note that there are only 5 lines total (4 in attr.h, 1 in cast.h) that this effort might improve, if successful.

`Debian clang version 11.0.1-21`
```
/usr/local/google/home/rwgk/forked/pybind11/include/pybind11/detail/../attr.h:534:9: warning: expression result unused [-Wunused-value]
        (int []) { 0, (process_attribute<typename std::decay<Args>::type>::precall(call), 0) ... };
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Reply:
I'll try to make use of what I just learned found out by introducing another macro, that does the warning suppression only for MSVC and clang.

Reply:
Done. Holding my breath that all compilers in our CI are happy with this.

It's a bit on the overkill side of the spectrum, but the comments that go with the new macro clearly document why some sort of workaround is needed.

Note that the code in the last commit is `clang-format`ed.

Reply:
Not even a flake, hooray!
@henryiii Please take another look.